core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/deploysemantics/PhysicalOp.scala

```diff
@@ -398,6 +440,31 @@ case class PhysicalOp(
     processingOrder.toList
   }
 
+  def propagateSchema(newInputSchema: Option[(PortIdentity, Schema)] = None): PhysicalOp = {
```

Comment:
Since we have `withInputSchema`, I think we can explicitly call `withInputSchema` before `propagateSchema` and remove line 444-448.

Reply:
I suggest we keep it this way to make propagateSchema the only API to be called from caller. This is to avoid caller's burden of managing withInputSchema, propagate, and withOutputSchema.